pkg/scm/git/git.go

```diff
@@ -62,18 +78,199 @@ func stringInSlice(s string, slice []string) bool {
+	}
+	// don't modify the protocol options which always have to be explicitly specified (http, https,  git)
+	if uri.Scheme != "ssh" && uri.Scheme != "file" {
+		return nil
+	}
+
+	// update the either file or ssh url accordingly
+	if len(mods.user) > 0 {
+		uri.User = url.User(mods.user)
+	}
+	if len(mods.scheme) > 0 {
+		uri.Scheme = mods.scheme
+	}
+	if len(mods.host) > 0 {
+		uri.Host = mods.host
+	}
+	if len(mods.path) > 0 {
+		uri.Path = mods.path
+	}
+	if len(mods.ref) > 0 {
+		uri.Fragment = mods.ref
+	}
+	return nil
+}
+
+// schemeNecessitatesReturn examines the URL and determines whether further
+// url / git spec string processing is needed or not, and if no more processing is needed,
+// whether the input should be deemed valid or not;
+// the combos of interest are:
+// - if a valid protocol/scheme is specifed, the caller should just return immediately (leaving the valid URL alone)
+// - if an invalid protocol/scheme is specified, just return immediately, but react appropriately to the fact the URL is invalid
+// - if no protocol/scheme is specified, don't return and continue trying to construct a valid URL
+func schemeNecessitatesReturn(uri *url.URL) (justReturn, invalidProtocol bool) {
+	if stringInSlice(uri.Scheme, allowedSchemes) {
+		justReturn = true
+		invalidProtocol = false
+		return
+	}
+
+	//at this point, the only valid value for Scheme should be "";  if it is
+	// set to some other protocol besides what is listed in allowedSchemes,
+	// say ftp://, then we return false, as it is a protocol we are not expecting.
+	// note, there also appears to be a bug in url.Parse, where a url like
+	// linux-box:/path/to/code will report the scheme as "linux-box"
+	if uri.Scheme != "" {
+		justReturn = true
+		invalidProtocol = true
+		return
+	}
+
+	justReturn = false
+	invalidProtocol = false
+	return
+}
+
+//func parseSpec(source string, remoteOnly bool) (validURL bool, scheme, user, host, path, ref string) {
+func parseSpec(source string, remoteOnly bool) (*urlMods, error) {
+	uri, err := url.Parse(source)
 	if err != nil {
-		return false
+		return nil, err
+	}
+
+	// to work around the url.Parse bug, deal with the fact it seemed to split on just :
+	if strings.Index(source, "://") == -1 && uri.Scheme != "" {
+		uri.Scheme = ""
+	}
+
+	if uri.Scheme == "file" && remoteOnly {
+		return nil, errors.New("file protocol is not a remote only spec")
+	}
+
+	justReturn, invalidProtocol := schemeNecessitatesReturn(uri)
+	if invalidProtocol {
+		return nil, errors.New(fmt.Sprintf("unsupported protocol specfied:  %s", uri.Scheme))
 	}
+	if justReturn {
+		return nil, nil
+	}
+
+	lastColonIdx := strings.LastIndex(source, ":")
+	atSignPresent := strings.Index(source, "@") != -1
+	// deal with the special git case where a user is not specified (i.e. no @);
+	// to do so, see if the entire spec as is exists as a file
+	trySSH := false
+	if lastColonIdx != -1 && !atSignPresent {
+		path := source
+		hashLoc := strings.LastIndex(source, "#")
```

Comment:
filenames can contain # too.  i'd just check the source exactly as is to see if it's a valid local path.